### DISCUSSION TITLE
[portraits] How to import models into spine2d

### DISCUSSION THREAD

**Discussion by StayFantasy** [nORbDragon/e7vault-issues#20] [General]
Hello I have a problem. I dwonloaded zklm's file and tried to import the data into the spine2d, but there was a problem, information such as :
```
ERROR: Error importing skeleton data file: /Users/e7herder/c1071/c1071.json
Please be sure to choose a valid skeleton JSON or binary file.
[error] Error reading file: /Users/e7herder/c1071/c1071.json
   at s.VfM._(_:178)
   at s.oIu._(_:323)
   at s.oIu.F(_:286)
   at s.KKJ._(_:1062)
   at s.oIu.C(_:279)
   at s.OOM.F(_:362)
   at s.GcY._(_:91)
   at s.gBm._(_:28)
   at s.EPy._(_:188)
   at s.EPy._(_:158)
   at com.badlogic.gdx.scenes.scene2d.ui.Button.F(_:125)
   at s.DGf._(_:93)
   at s.cbS._(_:88)
   at s.pMX._(_:71)
   at s.VgF.F(_:354)
   at s.AUx.F(_:124)
   at s.oXJ.F(_:75)
   at s.GAR._(_:75)
   at s.XJm._(_:117)
   at s.sZx.F(_:228)
   at s.BQU.F(_:138)
   at s.LMA.run(_:198)
   <events>
Cause: s.oMi: Error reading attachment: Layer 100, skin: default
   at s.XXQ._(_:312)
   at s.gxz._(_:118)
   at s.VfM._(_:150) ...
Cause: [arg] No enum constant s.Bsf.skinnedmesh
   at java.lang.Enum.valueOf(Unknown Source)
   at s.Bsf.valueOf(_:32)
   at s.XXQ._(_:370)
   at s.XXQ._(_:309) ...
```

I also tried to import json with model data,but there will be an error,like this:
```
ERROR: Error importing skeleton data file: /Users/e7herder/melissa/melissa.json
Please be sure to choose a valid skeleton JSON or binary file.
[error] Error reading file: /Users/e7herder/melissa/melissa.json
   at s.VfM._(_:178)
   at s.oIu._(_:323)
   at s.oIu.F(_:286)
   at s.KKJ._(_:1062)
   at s.oIu.C(_:279)
   at s.OOM.F(_:362)
   at s.GcY._(_:91)
   at s.gBm._(_:28)
   at s.EPy._(_:188)
   at s.EPy._(_:158)
   at com.badlogic.gdx.scenes.scene2d.ui.Button.F(_:125)
   at s.DGf._(_:93)
   at s.cbS._(_:88)
   at s.pMX._(_:71)
   at s.VgF.F(_:354)
   at s.AUx.F(_:124)
   at s.oXJ.F(_:75)
   at s.GAR._(_:75)
   at s.XJm._(_:117)
   at s.sZx.F(_:228)
   at s.BQU.F(_:138)
   at s.LMA.run(_:198)
   <events>
Cause: s.oMi: Error reading attachment: hair, skin: default
   at s.XXQ._(_:312)
   at s.gxz._(_:118)
   at s.VfM._(_:150) ...
Cause: [aioob] Index 50 out of bounds for length 50
   at s.XXQ._(_:510)
   at s.XXQ._(_:424)
   at s.XXQ._(_:309) ...
```
Could you give me some advice? Thanks.

**Comment by nORbDragon** [comment]
spine2d? You found spine v2.1.27? Or it is name for some modules for engines like unity? Or it is just other name for Spine v.3.x/v.4.x?

If you have Spine **v2.1.27**, I would "not mind" receiving it as a gift. =) It would be useful for me for my games and perversions.

Files from zklm's e7herder in something like "custom" format, does not fully comply with the official format for spine v2.1.27.

Some info about portaits and models (combat sprites) in this game:

* **portraits** and **combat sprites** created using EsotericSoftware Spine, but saved in a format convenient for the developers of this game of their **custom** design;
* **combat sprites** were created using Spine v2.1.27, **which is not found on the Internet**;
* **combat sprites** for each character are scattered over a bunch of files with additional information about how and with what additional inserts to animate them (S2 / S3, etc), I have not figured it out yet;
* **portraits** were also initially created using Spine v2.1.27, but from a certain moment they were drawn in Spine v3.x (different versions) using additional features from version 3;
* I have not yet fully analyzed the **custom format** of the developers of this game, how to convert their **custom** interpretation of Spine v3.x into original files for Spine v3.x;
* files from Spine v2.1.27 **cannot** be "just grabbed and opened" in Spine v3.x/v4.x, **even if you have a real license** (not a "torrent edition");
* you **can't** "just take and convert" Spine v2.1.27 files to Spine v3.x/v4.x, since critical functionality (FlipX/FlipY) was removed starting from Spine v3.

**zklm** dropped his work on e7herder when epic7 was moved from spine v.2.1.27 to a new format (v.3.x). All portraits on e7herder was created in spine v2.1.27, but saved in a "custom" format that is partially different from the original.

At the moment, I spend my free time on a full-fledged converter from Spine v2.1.27 to Spine v3.8.75. I'm looking for a way to imitate the removed functionality (FlipX/FlipY). To make it easier to deal with the game's combat sprites using Spine v3.x, and not waste my time catching all the bugs of the old engine from **e7herder**. Yes, there were bugs.

My early work on combat sprites (with glitches and bugs, spine v.2.1.27) - [https://www.norbdragon.com/e7vault/model-viewer.html?id=ras](https://www.norbdragon.com/e7vault/model-viewer.html?id=ras)

**Comment by StayFantasy** [comment]
Oh my god,  this model viewer is very good!!

Thanks for your reply.
My spine2d software was bought from the official website, it cost 329 USD, and then you buy the professional version, you can choose the version you want to open like this
![spine2d_Opening](https://user-images.githubusercontent.com/13538461/201306092-1a115542-2fdf-4928-bd49-af8f0ff87193.png)

If I use version 2.1.27 there is no detailed error message, like this
![spine2d_2_1_27](https://user-images.githubusercontent.com/13538461/201308526-3a888a01-07bb-44e6-a177-e87ff0705b93.png)

I wanted to directly import the json file and it should be displayed on the software, but it seems that I can't do this.
Maybe I should make the pictures on the portrait viewer work in the spine2d software.

Hope my information can help you.

**Comment by nORbDragon** [comment]
> My spine2d software

So, it is just Spine. =) I had doubts, maybe there is some other program with a similar name.

I have license for Spine too, but not "pro". My story about this - https://old.reddit.com/r/EpicSeven/comments/rfpz3x/one_common_decoder_day/


Hmm. But you have license for "pro", so you can try to import... Try import this (in spine v2.1.27 mode) - [c1071.zip](https://github.com/nORbDragon/e7vault-issues/files/9989426/c1071.zip)

I have been looking for a long time for someone who has a **professional edition (license, not torrent-edition)** who could try to import my work by switching to version 2.1.27.

**Comment by nORbDragon** [comment]
If you don't know how to unpack textures, use this archive - [c1071-w-images.zip](https://github.com/nORbDragon/e7vault-issues/files/9989473/c1071-w-images.zip)

**Comment by StayFantasy** [comment]
Hello I use [c1071-w-images.zip] this file,and result like this:
![螢幕快照 2022-11-11 下午9 33 04](https://user-images.githubusercontent.com/13538461/201351117-43caaee9-e66b-45bd-9f35-5a3cbd92b7c8.png)
The error message shows that something is missing,and the result after importing the image. like this:
![螢幕快照 2022-11-11 下午9 33 34](https://user-images.githubusercontent.com/13538461/201351424-e668db71-47c0-4f07-8ef7-0221d51c6d5f.png)

add face picture
![螢幕快照 2022-11-11 下午10 17 52](https://user-images.githubusercontent.com/13538461/201359051-6ba74bf9-efc2-46ca-9fe7-d018aa1fe535.png)

**Comment by nORbDragon** [comment]
`Nonessential not checked when exporting ...` <- this is normal warning when you import from json-file.

What we have missing on your second screenshot: it is links "slot" <-> "image". Something wrong. Well, even that result is not bad for me. Json-file is IMPORTED. Finaly.

I will contact with you in next monday with new variant for spine v2.1.27.

This is how she must be seen in spine (I still working on converter spine v2 > spine v3):

![Screenshot 2022-11-11 171611](https://user-images.githubusercontent.com/97672704/201358526-9eab21b3-319f-4c17-9f06-30d88cdf645a.png)

BTW, open in right panel **root** > **center** > **bone156** > **leg_l_02**. I'm wondering what it shows in the slot instead of the sprite attached to this slot. =)

**Comment by StayFantasy** [comment]
root > center > bone156 > leg_l_02 , Is it like this?
![螢幕快照 2022-11-11 下午10 29 24](https://user-images.githubusercontent.com/13538461/201361288-97167483-d476-467f-9505-a0160d1910a8.png)

**Comment by nORbDragon** [comment]
Err... Image for leg is here. Hmm.

Can you close "Outline" window? I need to see ALL little windows on screen's bottom.

**Comment by StayFantasy** [comment]
![螢幕快照 2022-11-11 下午10 37 26](https://user-images.githubusercontent.com/13538461/201362543-58d922f4-7914-47e0-aa8e-05422d51b01d.png)

edit mesh mode
![螢幕快照 2022-11-12 下午3 01 06](https://user-images.githubusercontent.com/13538461/201462242-165616c1-9bf5-41f7-b837-bacac3e4bcc7.png)

**Comment by nORbDragon** [comment]
Strange. Need to think about that and compare examples from spine v2 and my json-file. What I missed.

I will contact you in next monday.

**Comment by StayFantasy** [comment]
OK, thank you for your help.

**Comment by nORbDragon** [comment]
Try this - [c1071-json.zip](https://github.com/nORbDragon/e7vault-issues/files/10001124/c1071-json.zip), replace old json-file with this file.

Little changes in leg (leg_l_02). Any changes there? Screenshot please. =)

**Comment by StayFantasy** [comment]
result :
![螢幕快照 2022-11-14 下午5 53 09](https://user-images.githubusercontent.com/13538461/201631734-f47335f1-3056-45df-af48-98ba0b3b7382.png)

edit mesh mode
![螢幕快照 2022-11-14 下午5 59 40](https://user-images.githubusercontent.com/13538461/201631815-419d6eee-5275-4bfd-8c7c-d2aefd02d708.png)

It looks like the mesh is too small to match the picture.

**Comment by nORbDragon** [comment]
When you import, Spine doesn't ask you any questions? Like this:

![Screenshot 2022-11-14 133835](https://user-images.githubusercontent.com/97672704/201639898-07bc33ae-a915-4cfa-9c53-b4216ae162a3.jpg)

**Comment by StayFantasy** [comment]
NO. my message is like this
![201351117-43caaee9-e66b-45bd-9f35-5a3cbd92b7c8 (1)](https://user-images.githubusercontent.com/13538461/201644873-9c6fa232-5b4f-481f-a471-cd6d83eb5542.png)

Maybe the spine 2.1.27 version does not have such a function

**Comment by nORbDragon** [comment]
Wait. You have two sceletons? **"c1071"** and **"sceleton"**. How you do import?

I import by **"menu"** > **"Import Data..."** > select json-file + **check** on **"New project"** > **"Import"** button > **"Discard changes"** > ok for **"Import warning"** > no for **"Images Resized"**

![Screenshot 2022-11-14 141709](https://user-images.githubusercontent.com/97672704/201647746-08df2506-3e14-4d1e-b0be-5e3419737549.jpg)
![Screenshot 2022-11-14 141733](https://user-images.githubusercontent.com/97672704/201647752-f7dead3b-5438-47f8-b2ca-5ff7e795f7c3.jpg)
![Screenshot 2022-11-14 141750](https://user-images.githubusercontent.com/97672704/201647757-91cfa09e-8e95-4491-9a6c-596b8b258d2f.jpg)
![Screenshot 2022-11-14 141811](https://user-images.githubusercontent.com/97672704/201647759-e62f5e49-edb0-4968-8805-53d0e30fd330.jpg)
![Screenshot 2022-11-14 141827](https://user-images.githubusercontent.com/97672704/201647764-b28bfe8a-ff01-4d90-a301-f76c1b3e62bc.jpg)
![Screenshot 2022-11-14 141842](https://user-images.githubusercontent.com/97672704/201647766-98449ad8-47ed-4852-96ed-a32bf293383a.jpg)
![Screenshot 2022-11-14 142625](https://user-images.githubusercontent.com/97672704/201648599-40350249-8c49-4cc2-8fbd-9841cf3b37fd.jpg)


How you can see, I don't have two sceletons on right panel, only one - **c1071**.

How I can see on your screens, you have two sceletons. And "Images" was connected to **"sceleton"**, not to **"c1071"** sceleton. Something wrong there.

Can you just delete "sceleton" and import again?

**Comment by StayFantasy** [comment]
In the video this is how I import the file
https://drive.google.com/file/d/1I2z3dumg7NTcXvJZO__iDJkE4FHsN5P2/view?usp=sharing

**Comment by nORbDragon** [comment]
Hmm. MacOS? Maybe this variant will work - [c1071-json-02.zip](https://github.com/nORbDragon/e7vault-issues/files/10002827/c1071-json-02.zip)

I expect it should now automatically pick up the images folder. No need to manually specify it.

**Comment by StayFantasy** [comment]
I use macOS. the result still doesn't work
![螢幕快照 2022-11-14 下午8 31 40](https://user-images.githubusercontent.com/13538461/201660919-506df5f0-79f1-4686-a43c-6769d15713a0.png)

**Comment by nORbDragon** [comment]
But **Images** folder picked up automatically now? Or not?

Okay. Let's try from the other end. 

Two versions with a leg. The rest was thrown out of the file so as not to interfere. Let's get on the leg. 

It is possible that the second variant is not imported, will give an error. This will also be useful information for me.

[c1071-leg-01.zip](https://github.com/nORbDragon/e7vault-issues/files/10003173/c1071-leg-01.zip) / [c1071-leg-02.zip](https://github.com/nORbDragon/e7vault-issues/files/10003174/c1071-leg-02.zip)

**Comment by nORbDragon** [comment]
And variant 3, try this too: [c1071-leg-03.zip](https://github.com/nORbDragon/e7vault-issues/files/10003195/c1071-leg-03.zip)

**Comment by nORbDragon** [comment]
@StayFantasy 

Since we are discussing a **"issue"** that in no way relates to my site, I decided to move it from **"issues"** to regular **"discussions"**.

**Comment by StayFantasy** [comment]
Hello, this is result: 

c1071-leg-01.zip
https://user-images.githubusercontent.com/13538461/201808949-13a2ec40-2fa5-4b53-8ad8-12242ed324be.mov

c1071-leg-02.zip
https://user-images.githubusercontent.com/13538461/201809134-74fc6085-a084-420d-a8fe-629b440162ed.mov

c1071-leg-03.zip
https://user-images.githubusercontent.com/13538461/201809234-1e7db755-fa88-4197-934b-ee9e93acfde2.mov

I have some ideas, would you like to give me your project file so I can open it?

**Comment by nORbDragon** [comment]
There are no such files like *"project file"* in the game client.

There are atlas-, sct- and scsp-files. 

- atlas-files from game client is same as atlas-files from spine;
- sct-files is custom format for images, can be converted to png-files;
- scsp-files is custom binary format like skel (binary export format from spine).

If you are interested in how the original files look from the game client - [portrait-c1071-original.zip](https://github.com/nORbDragon/e7vault-issues/files/10009955/portrait-c1071-original.zip)

I am converting **scsp**-file to **json**-format on my own. I don't have spine v2.1.27. So I made the converter blindly, debugging by using e7herder engine, not by Spine.

Later I created converter scsp to spine v2.1.27 by testing on old skeletonViewer.jar - [skeletonviewer-spine2.zip](https://github.com/nORbDragon/e7vault-issues/files/10010028/skeletonviewer-spine2.zip). No bugs was there, so I can't understand why we have troubles in your spine v2.1.27.

I still think there is something wrong with image folder. I can't understand why spine does not want to automatically connect to the folder with pictures. 

My silly assumption: due to the inability to find pictures, spine resets some important settings for their connection with bones/slots/meshes.

Can you just move folder with my files from "user folder" to other place? Maybe something wrong with user/application rights for access to user's private folder.

And you can play with my json-file. Change value for "images" to real path to forder with images. Maybe MacOs can't work with relative path.

**Comment by nORbDragon** [comment]
BTW: you can try remove *"skeleton">"images"* line from json-file and move files from "images" folder to parent folder, where **c1071.json** is. Or just try this - [c1071-20221115.zip](https://github.com/nORbDragon/e7vault-issues/files/10010756/c1071-20221115.zip)

**Comment by StayFantasy** [comment]
well...When I use the 2.1 version to save as a project file, and then select the 4.1.17 version to open the project, there is no problem.
https://drive.google.com/file/d/1GhLS6-3f4bbLrb3LrO9JaeiPgG9H-RM9/view?usp=sharing
https://drive.google.com/file/d/1knG1fL5y1IMTbzCF40duPa62D1qy7H-g/view?usp=sharing

**Comment by nORbDragon** [comment]
... no comments. 

Can you load your project from 2.1.27 in... **2.1.27**? Yep, in same version.

Maybe v4.1.17 is not need for fix. And just save/load project need.

Some portraits use FlipX/FlipY from v2.1.27, so you can't just save them in 2.1.27 and open in 3.x/4.x versions.

**Reply by StayFantasy** [reply]
Hello, sorry for the long time no message.
I have some troubles in life, and it will delay spine2d.
If I have new discoveries, I will share them with you.

**Comment by nORbDragon** [comment]
@StayFantasy , can you help me? =)

Can you:
- open [c3052-spine2.zip](https://github.com/nORbDragon/e7vault-issues/files/10425206/c3052-spine2.zip) in spine v2.1.27;
- save this as project file;
- open project file in spine v3.8;
- export to json in v3.8 format.

Something strange with Jena's staff angle when I trying convert this portrait from v2.1.27 to v3.8 by myself. And I can't understand why.

I need json from spine v3.8 for compare.

**Reply by StayFantasy** [reply]
Hi This is the json file for version 3.8
[version3-8.zip](https://github.com/nORbDragon/e7vault-issues/files/10425820/version3-8.zip)
By the way, this is the screen on my 3.8 version
![version3-8_view](https://user-images.githubusercontent.com/13538461/212680418-c1b68caf-6446-4923-86af-189a0d82a00f.png)
hope this helps.

**Comment by nORbDragon** [comment]
Yeah. So, same thing with right arm / staff's angle after conversion by spine program.

Thank you for experiment. =)

Original (spine v2):
![spine-2-original](https://user-images.githubusercontent.com/97672704/212687896-38b0b882-72e5-468e-8191-1e453d650707.jpg)

I circled in red what I don't like:

![spine-3-official](https://user-images.githubusercontent.com/97672704/212687908-d8d323fa-d538-4507-8a57-04ae10cb763c.jpg)

your variant after convertion by spine ^

![spine-3-my-experiments](https://user-images.githubusercontent.com/97672704/212687900-98b50d50-39ef-4102-a720-97190a2bb952.jpg)

my variant after convertion by my scripts ^ (played with some values, so do not pay attention to the not quite correct aspect ratio for the pommel of the staff)

**Comment by nORbDragon** [comment]
@StayFantasy , maybe you can figure out how this can be.

Download this (json for spine v3) - [c3052-spine3.zip](https://github.com/nORbDragon/e7vault-issues/files/10466363/c3052-spine3.zip)

Import data and try do some changes there:
- find bone "wrist";
- change "Axes" to "World" mode;
- change "Rotate" value to "212";
- press enter for apply changes.

My spine editor "autofix" rotate value from 212 to 217.331. I can't understand why. Maybe you have some ideas?


![spine-01](https://user-images.githubusercontent.com/97672704/213697701-ae957518-ded3-4a4c-88f9-b9d5f68af18a.jpg)
![spine-02](https://user-images.githubusercontent.com/97672704/213697715-5818f9d6-0ac8-4bc5-a9ec-8aa07ccf2cb9.jpg)
![spine-03](https://user-images.githubusercontent.com/97672704/213697727-8617e58a-3d57-4b83-b15d-006527befe69.jpg)

**Reply by StayFantasy** [reply]
Hello, the computer is not with me at the moment, it will take two weeks before I can help you.

**Reply by nORbDragon** [reply]
Ouch. No problems. =)

**Comment by StayFantasy** [comment]
Hello, I did some testing, and I found that the values entered seem to be a little different, but the values scrolled with the mouse wheel feel more correct.

If you scroll to the 212 value, you will find that it is not the same as the 212 value you entered.

keyboard :
https://drive.google.com/file/d/1-Fp8j-2EmmgSC1fo0_xo7wEGKHokBinS/view

Mouse :
https://drive.google.com/file/d/11R7y8BECYTyzgHCYRW8E91z-NLOsURUV/view

**Reply by nORbDragon** [reply]
I don't understand why version 3.x breaks bone scaling and rotation.

Found a way around this problem through the use of transform constraint. So far I have not done a script for automatically adding transform constraint, but the result looks quite good (imho). 

just json with transform constraint - [c3052_w_transform.zip](https://github.com/nORbDragon/e7vault-issues/files/10576495/c3052_w_transform.zip)

The animation still doesn't quite match the original. I think somewhere else the bones are scaled and rotated incorrectly by spine v3.x engine, and they need the same neat fix.

**[EDITED]:** set **"idle"** animation for check in spine, not "idle2"/"idle3".
